irohad/ordering/impl/on_demand_ordering_gate.cpp

```diff
@@ -0,0 +1,75 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "ordering/impl/on_demand_ordering_gate.hpp"
+
+#include "interfaces/iroha_internal/proposal.hpp"
+#include "interfaces/iroha_internal/transaction_batch.hpp"
+
+using namespace iroha;
+using namespace iroha::ordering;
+
+OnDemandOrderingGate::OnDemandOrderingGate(
+    std::shared_ptr<OnDemandOrderingService> ordering_service,
+    std::shared_ptr<transport::OdOsNotification> network_client,
+    rxcpp::observable<BlockRoundEventType> events,
+    std::unique_ptr<shared_model::interface::UnsafeProposalFactory> factory,
+    transport::RoundType initial_round)
+    : ordering_service_(std::move(ordering_service)),
+      network_client_(std::move(network_client)),
+      events_subscription_(events.subscribe([this](auto event) {
+        // exclusive lock
+        std::lock_guard<std::shared_timed_mutex> lock(mutex_);
+
+        visit_in_place(event,
+                       [this](BlockEvent block) {
+                         // block committed, increment block round
+                         current_round_ = {block.height, 1};
+                       },
+                       [this](EmptyEvent empty) {
+                         // no blocks committed, increment reject round
+                         current_round_ = {current_round_.first,
+                                           current_round_.second + 1};
+                       });
+
+        // notify our ordering service about new round
+        ordering_service_->onCollaborationOutcome(current_round_);
+
+        // request proposal for the current round
+        auto proposal = network_client_->onRequestProposal(current_round_);
+
+        // vote for the object received from the network
+        proposal_notifier_.get_subscriber().on_next(
+            std::move(proposal).value_or_eval([&] {
+              return proposal_factory_->unsafeCreateProposal(
+                  current_round_.first, current_round_.second, {});
+            }));
+      })),
+      proposal_factory_(std::move(factory)),
+      current_round_(initial_round) {}
+
+void OnDemandOrderingGate::propagateTransaction(
+    std::shared_ptr<const shared_model::interface::Transaction> transaction)
+    const {
+  throw std::logic_error("Not implemented");
```

Comment:
Maybe write "deprecated" instead of "logic error"?

Reply:
I miss with that. I mean "deprecated" instead of "Not implemented". Sorry about that.